### PR TITLE
[BIM-Webviewer] More detail on Webviewer Events

### PR DESCRIPTION
**Pull request by sbieserprocore**
This PR gives a bit more detail about the Events Namespace portion of the Webviewer API.

* Removed event names table
* Added new Event Names section, with details about each event and detail about return data for callback functions
* Events Namespace section now only contains the methods.
* Fixed some typos and wordings in other sections.
* Added code examples
* Documented 8 more events and removed a few that were never firing or hooked up in the webviewer.

UPDATED 8/25/2022 2:00pm:
[events_update.pdf](https://github.com/procore/documentation/files/9428238/events_update.pdf)

| Before | After |
| ----------- | ----------- |
|image|image|

### CODE REVIEW

**Comment by sbieserprocore** [comment]
Need to add new events introduced by https://github.com/procore/bim-webviewer/pull/590

File: other_procore_apis/bim_web_viewer.md [hunk 1]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+
+---
+
+### coachmarkHyperlinkClicked
+
+<p class="heading-link-container"><a class="heading-link" href="#coachmarkhyperlinkclicked"></a></p>
+
+Fires when the coachmark indicator has been cleared. The coachmark in the upper right hand corner of the webviewer appears due to; hidden objects, applied section planes, or a field of view has been applied other than the default field of view. When you clear a coachmark, this event fires and returns a string value of which coachmark was cleared.
+
+This string can be one of the following values: `hidden`, `section`, or `fov`.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:888]: I found this a little hard to parse. I tried rewriting in my native language of bullet points 🚄 . Please consider :)

```suggestion
Fires when the coachmark indicator has been cleared. The coachmark will appear in the upper right hand corner of the webviewer appear when:
- objects are hidden
- section planes are applied
- a field of view has been applied other than the default field of view

When the coachmark is cleared, this event will fire with a value indicating which type of coachmark was cleared:
- `"hidden"`
- `"section"`
- `"fov"`
```

File: other_procore_apis/bim_web_viewer.md [hunk 2]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+
+Fires after a double click occurs on the webviewer container element. Returns a [MouseEvent](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent) object.
+
+---
+
+### downloadComplete
+
+<p class="heading-link-container"><a class="heading-link" href="#downloadcomplete"></a></p>
+
+Fires after all the requests for the model has been downloaded. This event will still fire even if the model is cached.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:904]:
```suggestion
Fires after all download requests for the model have been completed. This event will still fire even if the model is cached.
```

File: other_procore_apis/bim_web_viewer.md [hunk 3]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+
+Fires before requests are made to download the model.
+
+---
+
+### drawingNavigated
+
+<p class="heading-link-container"><a class="heading-link" href="#drawingnavigated"></a></p>
+
+Fires when the 2D Navigation overlay is used to navigate the model. Specifically, this fires after a location and direction has been chosen on the floor plan and the overlay is dismissed.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:936]:
```suggestion
Fires when the 2D Navigation overlay is used to navigate the model. Specifically, this fires after a location and direction have been chosen on the floor plan and the overlay is dismissed.
```

File: other_procore_apis/bim_web_viewer.md [hunk 4]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+
+<p class="heading-link-container"><a class="heading-link" href="#navigationchanged"></a></p>
+
+Fires when the navigation mode has changed.
+
+#### Data Properties
+
+| Field Name | Type   | Description                                     |
+| ---------- | ------ | ----------------------------------------------- |
+| mode       | number | The type of navigation that was just activated. |
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1139]: Guessing we're not documenting this specifically because it just gives you 0, 1, 2, 3 ? haha

I think that's coming from here and we don't expose it currently: https://github.com/procore/bim-webviewer/blob/develop/src/controllers/RenderWrapperController.js#L156-L162

File: other_procore_apis/bim_web_viewer.md [hunk 5]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+| controls   | object | An instance of the controller object.           |
+
+---
+
+### objectDoubleClick
+
+<p class="heading-link-container"><a class="heading-link" href="#objectdoubleclick"></a></p>
+
+Fires on double mouse clicks. If the double mouse click occurs on an object, this event returns a Meshnode ID, otherwise -1 is returned. A double click also produces one `objectSingleClick` event.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1148]: I think where the SDK API audit landed was that we would call these meshnode indices. 

```suggestion
Fires on double mouse clicks. If the double mouse click occurs on an object, this event returns a Meshnode Index, otherwise -1 is returned. A double click also produces one `objectSingleClick` event.
```

File: other_procore_apis/bim_web_viewer.md [hunk 6]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+
+---
+
+### objectSingleClick
+
+<p class="heading-link-container"><a class="heading-link" href="#objectsingleclick"></a></p>
+
+Fires on a single mouse click. If the single mouse click occurs on an object, this event returns Meshnode ID, otherwise -1 is returned.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1156]:
```suggestion
Fires on a single mouse click. If the single mouse click occurs on an object, this event returns a Meshnode Index, otherwise -1 is returned.
```

File: other_procore_apis/bim_web_viewer.md [hunk 7]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+| ---------- | ------ | --------------------------------------------------------------------------------------------------------- |
+| origin     | string | The origin of the un hide operation, can be: `coachmark_hidden`, `object_tree`, `object_tree_all_objects` |
+
+---
+
+### objectSelect
+
+<p class="heading-link-container"><a class="heading-link" href="#objectselect"></a></p>
+
+Fires when an object has been selected. This event returns Meshnode ID.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1192]:
```suggestion
Fires when an object has been selected. This event returns a Meshnode Index.
```

File: other_procore_apis/bim_web_viewer.md [hunk 8]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+
+Fires when the Webviewer options have changed. This event fires when `setOptions` is called, or when actions cause the multiple selection option to change, such as keyboard shortcuts to multi select objects.
+
+---
+
+### objectMeasure
+
+<p class="heading-link-container"><a class="heading-link" href="#objectmeasure"></a></p>
+
+Fires objects are selected for measurement. This event returns a Meshnode ID of the object selected fore measurement.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1208]:
```suggestion
Fires when objects are selected for measurement. This event returns a Meshnode Index of the object selected for measurement.
```

File: other_procore_apis/bim_web_viewer.md [hunk 9]
```diff
@@ -832,13 +791,541 @@ Manually dispatches an event
...
+<p class="heading-link-container"><a class="heading-link" href="#selecttool"></a></p>
+
+Fires when a tool has been activated.
+
+#### Data Properties
+
+| Field Name | Type   | Description                                                                                 |
+| ---------- | ------ | ------------------------------------------------------------------------------------------- |
+| origin     | string | The origin of the tool being activated.                                                     |
+| tool       | number | The tool that has been activated, enumerated from `ProcoreBim.Webviewer.toolstoolbar_enum`. |
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1248]: Looks like this should be `toolbar_enum`.

image


```suggestion
| tool       | number | The tool that has been activated, enumerated from `ProcoreBim.Webviewer.toolbar_enum`. |
```

File: other_procore_apis/bim_web_viewer.md [hunk 10]
```diff
@@ -180,7 +180,7 @@ None
 ##### Returns
 
 ```js
-undefined
+undefined;
```
[the remark this answers is not in the document]
Comment (sbieserprocore) [reply other_procore_apis/bim_web_viewer.md:183]: Let me check out the prettier, it ultimately made these changes with the semicolon and the table spacings.

File: other_procore_apis/bim_web_viewer.md [hunk 11]
```diff
@@ -507,14 +509,14 @@ Moves camera from current position to fit the axis-aligned bounding box of the n
 
 #### Parameters
 
-| Field Name | Required | Type | Description |
-| - | - | - | - |
-| objectIds | true | ObjectId[] | An array of object ids. To zoom to a single object, pass an array of one. |
+| Field Name | Required | Type       | Description                                                               |
+| ---------- | -------- | ---------- | ------------------------------------------------------------------------- |
+| objectIds  | true     | ObjectId[] | An array of object ids. To zoom to a single object, pass an array of one. |
```
[the remark this answers is not in the document]
Comment (sbieserprocore) [reply other_procore_apis/bim_web_viewer.md:514]: Same comment: https://github.com/procore/documentation/pull/102#discussion_r954378908

I will check out the prettier.

File: other_procore_apis/bim_web_viewer.md (hunk 4 above)
Comment (sbieserprocore) [reply other_procore_apis/bim_web_viewer.md:1145]: Yup, that is correct lol.

Comment (sbieserprocore) [reply other_procore_apis/bim_web_viewer.md:1145]: We can update documentation if we decide to expose that enum.

File: other_procore_apis/bim_web_viewer.md [hunk 12]
```diff
@@ -839,6 +796,540 @@ undefined
...
+// Perhaps your application places markers that represent 3D positions on the 2D floor plan overlay while it is open. When
+// tapping on a marker, this function could be called that uses `floorPlanNavigateAndClose` to set the camera and close the
+// 2D floor plan overlay.
+const goto = (camera) => {
+  if (floorPlanNavigateAndClose !== null) {
+    floorPlanNavigateAndClose(camera.x, camera.y, camera.z, 0.1, 0.1);
+    floorPlanNavigateAndClose = null;
+  }
+};
+```
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:1025]: This is maybe starting to get nitpicky, but I found it confusing that `floorPlanNavigateAndClose` is at the same level of hierarchy as `floorPlan`, which makes it look like an actual event as opposed to a parameter of the `floorPlan` callback. Maybe we could make `floorPlanNavigateAndClose` and `floorPlanPoint` (and the headers under them) one level deeper.

I'll leave individual line suggestions to make it clearer what I'm proposing to change.

File: other_procore_apis/bim_web_viewer.md [hunk 13]
```diff
@@ -839,6 +796,540 @@ undefined
...
+
+| Field Name | Required | Type   | Description                |
+| ---------- | -------- | ------ | -------------------------- |
+| x          | true     | number | x coordinate in 3D space.  |
+| y          | true     | number | y coordinate in 3D space.  |
+| z          | true     | number | z coordinate in 3D space.  |
+| dx         | true     | number | x direction of the camera. |
+| dy         | true     | number | y direction of the camera. |
+
+### floorPlanPoint
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:983]:
```suggestion
#### floorPlanPoint
```

File: other_procore_apis/bim_web_viewer.md [hunk 14]
```diff
@@ -839,6 +796,540 @@ undefined
...
+
+<p class="heading-link-container"><a class="heading-link" href="#floorplanpoint"></a></p>
+
+Returns the corresponding point on the 2D floor plan overlay from a point in 3D model space.
+
+#### Parameters
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:989]:
```suggestion
##### Parameters
```

File: other_procore_apis/bim_web_viewer.md [hunk 15]
```diff
@@ -839,6 +796,540 @@ undefined
...
+| overlay                   | HTMLElement | The HTMLElement for the 2D Navigation overlay                                                               |
+| type                      | string      | The type of interaction and has the following values: `open`, `close`, and `update`                         |
+
+### floorPlanNavigateAndClose
+
+<p class="heading-link-container"><a class="heading-link" href="#floorplannavigateandclose"></a></p>
+
+Updates the camera position and direction in addition to dismissing the 2D floor plan overlay.
+
+#### Parameters
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:973]:
```suggestion
##### Parameters
```

File: other_procore_apis/bim_web_viewer.md [hunk 16]
```diff
@@ -839,6 +796,540 @@ undefined
...
+#### Data Properties
+
+| Field Name                | Type        | Description                                                                                                 |
+| ------------------------- | ----------- | ----------------------------------------------------------------------------------------------------------- |
+| floorPlanNavigateAndClose | function    | Provides a callback function to navigate the camera to a position and close the overlay. Returns a promise. |
+| floorPlanPoint            | function    | Provides a callback function to calculate a point on the current floor plan. Returns a promise.             |
+| overlay                   | HTMLElement | The HTMLElement for the 2D Navigation overlay                                                               |
+| type                      | string      | The type of interaction and has the following values: `open`, `close`, and `update`                         |
+
+### floorPlanNavigateAndClose
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:967]:
```suggestion
#### floorPlanNavigateAndClose
```

File: other_procore_apis/bim_web_viewer.md [hunk 17]
```diff
@@ -839,6 +796,540 @@ undefined
...
+
+| Field Name | Required | Type   | Description                 |
+| ---------- | -------- | ------ | --------------------------- |
+| x          | true     | number | x coordinate of the camera. |
+| y          | true     | number | y coordinate of the camera. |
+| z          | true     | number | z coordinate of the camera. |
+| dx         | true     | number | x direction of the camera.  |
+| dy         | true     | number | y direction of the camera.  |
+
+#### Example
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:999]: No change here. Children of `floorPlan` are:
- Data Properties
- `floorPlanNavigateAndClose`
- `floorPlanPoint`
- Example

Feels right to me.

File: other_procore_apis/bim_web_viewer.md [hunk 18]
```diff
@@ -839,6 +796,540 @@ undefined
...
+});
+
+viewer.start();
+```
+
+### Callback function
+
+<p class="heading-link-container"><a class="heading-link" href="#callback-function"></a></p>
+
+The second parameter to `addEventListener` is a third party defined function to be executed when the event fires. Depending on the event, the callback function may return undefined, or data. This data may be a number, string, or a JavaScript object.
```
Comment (Bates550) [inline other_procore_apis/bim_web_viewer.md:829]: Feels like this callback function was meant to go somewhere else? Maybe part of `addEventListener`

image

File: other_procore_apis/bim_web_viewer.md (hunk 12 above)
Comment (sbieserprocore) [reply other_procore_apis/bim_web_viewer.md:1025]: Yes! This is great. I was wondering about this too, I think this is a perfect solution to this.

File: other_procore_apis/bim_web_viewer.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (sbieserprocore) [reply other_procore_apis/bim_web_viewer.md]: I have opted to just revert these changes.

**Review by Bates550 — APPROVED** [review]
LGTM! Awesome work, thanks for making such a huge improvement to the docs 🥳
